plan/physical_plan_builder.go

```diff
@@ -783,6 +785,178 @@ func (p *Union) convert2PhysicalPlan(prop *requiredProperty) (*physicalPlanInfo,
+	}
+	for _, idx := range indices {
+		// TODO: Currently we don't consider composite index.
+		if len(idx.Columns) > 1 {
+			continue
+		}
+		checker := &conditionChecker{
+			idx:          idx,
+			columnOffset: 0,
+			length:       idx.Columns[0].Length,
+		}
+		// This idx column should occur in one condition which contains both column and correlated column.
+		// And conditionChecker.check(this condition) should be true.
+		var usable bool
+		for _, cond := range newConds {
+			// If one cond is ok, then this index is useful.
+			if checker.check(cond) {
+				usable = true
+				break
+			}
+		}
+		if usable {
+			usableIdxs = append(usableIdxs, idx)
+		}
+	}
+	var pkName model.CIStr
+	if ds.tableInfo.PKIsHandle {
+		var pkCol *expression.Column
+		for i, col := range ds.Columns {
+			if mysql.HasPriKeyFlag(col.Flag) {
+				pkCol = ds.schema.Columns[i]
+				break
+			}
+		}
+		if pkCol == nil {
+			return usableIdxs, pkName
+		}
+		checker := conditionChecker{
+			pkName: pkCol.ColName,
+			length: types.UnspecifiedLength,
+		}
+		// Pk should satisfies the same property as the index.
+		var usable bool
+		for _, cond := range newConds {
+			if checker.check(cond) {
+				usable = true
+				break
+			}
+		}
+		if usable {
+			pkName = pkCol.ColName
+		}
+	}
+	return usableIdxs, pkName
+}
+
+func (p *Selection) makeScanController() *physicalPlanInfo {
+	newSel := *p
+	newSel.ScanController = true
+	var child PhysicalPlan
+	ds := p.children[0].(*DataSource)
+	if p.usefulPkName.L != "" {
+		ts := &PhysicalTableScan{
+			Table:               ds.tableInfo,
+			Columns:             ds.Columns,
+			TableAsName:         ds.TableAsName,
+			DBName:              ds.DBName,
+			physicalTableSource: physicalTableSource{client: ds.ctx.GetClient()},
+		}
+		ts.tp = Tbl
+		ts.allocator = ds.allocator
+		ts.SetSchema(ds.schema)
+		ts.initIDAndContext(ds.ctx)
+		if ds.ctx.Txn() != nil {
+			ts.readOnly = p.ctx.Txn().IsReadOnly()
+		} else {
+			ts.readOnly = true
+		}
+		child = ts
+	} else {
+		var chosenPlan *PhysicalIndexScan
+		sc := ds.ctx.GetSessionVars().StmtCtx
+		client := ds.ctx.GetClient()
+		for _, idx := range p.usefulIndices {
+			condsBackUp := make([]expression.Expression, 0, len(p.Conditions))
+			for _, cond := range p.Conditions {
+				condsBackUp = append(condsBackUp, cond.Clone())
+			}
+			is := &PhysicalIndexScan{
+				Table:               ds.tableInfo,
+				Index:               idx,
+				Columns:             ds.Columns,
+				TableAsName:         ds.TableAsName,
+				OutOfOrder:          true,
+				DBName:              ds.DBName,
+				physicalTableSource: physicalTableSource{client: ds.ctx.GetClient()},
+			}
+			is.tp = Idx
+			is.allocator = ds.allocator
+			is.SetSchema(ds.schema)
+			is.initIDAndContext(ds.ctx)
+			if is.ctx.Txn() != nil {
+				is.readOnly = p.ctx.Txn().IsReadOnly()
+			} else {
+				is.readOnly = true
+			}
+			is.AccessCondition, condsBackUp = DetachIndexScanConditions(condsBackUp, is)
```

Comment:
Calculating access condition is to choose which index is better by accessEqualCount and accessInAndEqCount.
Idx conds and tbl conds is used to make explain result clear, can be removed.